### PR TITLE
add feature to remove graph from plot2d

### PR DESCRIPTION
**Pull request by vbargsten**
I'm unsure if this is the right way to disconnect and especially why the port has to be deleted from on_data_listeners completely. If it is kept, we will have multiple ports with the same name if the signal is plotted again after removing.

note: requires corresponding merge in gui-rock_widget_collection

### CODE REVIEW

**Comment by doudou** [comment]
@D-Alex if you had 5 minutes to have a quick look, I would appreciate it

**Comment by D-Alex** [comment]
It looks dangerous to stop and delete the port because as far as I remember the port is shared between all widgets. Therefore, other widgets might stop producing data. You would have to stop only the listener associated to the widget.

**Comment by vbargsten** [comment]
> It looks dangerous to stop and delete the port because as far as I remember the port is shared between all widgets.

The port is not stopped or deleted. It is _only_ deleted from the port<->listeners hash of the ConnectionManager  associated with the widget (I checked that each new widget gets its own ConnectionManager).  So we basically undo `connect_to`'s addition of port/listener to the hash ( `@on_data_listeners[port] << listener` ).

> You would have to stop only the listener associated to the widget.

That's what is done by calling the disconnect method with the respective port (before removing the port from the widget's connection manager )

**Comment by D-Alex** [comment]
👍

**Comment by doudou** [comment]
> That's what is done by calling the disconnect method with the respective port (before removing the port from the widget's connection manager )

As it is `#remove_port` is a fairly dangerous method. Please check first that there are no listeners associated with the port (and mention in the exception message that `#disconnect` needs to be called first)

**Comment by vbargsten** [comment]
I've made it an option of disconnect now. This makes sure all listeners were stopped before. Please check again.

File: lib/vizkit/connection_manager.rb [hunk 1]
```diff
@@ -48,17 +44,23 @@ def listening?(port = nil)
             end
         end
 
-        def disconnect(port=nil)
+        def disconnect(port=nil, keep_port=true)
```
Comment (doudou) [inline lib/vizkit/connection_manager.rb]: Please use a keyword argument here

```
def disconnect(port = nil, keep_port: true)
end
```

and `disconnect(port, keep_port: false)`

**Comment by doudou** [comment]
Thanks !

**Pull request merged**